webapp/controller/ListReport.controller.js

```diff
@@ -0,0 +1,302 @@
+				filters: this.getCategoriesFilters(),
+				and: false
+			})
+			
+			const oSuppliersFilters = new Filter({
+				filters: this.getSuppliersFilters(aSelectedId),
+				and: false
+			})
+			const oPriceFilter = new Filter({
+				filters: this.getPriceFilter(),
+				and: false
+			})
+			return new Filter({
+				filters: [oSearchFilter, oAvailabilityFilter, oCategoryFilters, oSuppliersFilters, oPriceFilter],
+				and: true
+			});
+		},
+
+		onFilter: function(aSelectedId = null) {
+			const oTable = this.byId("table");
+			const oItemsBinding = oTable.getBinding("items");
+			
+			const oFilter = this.getCombinedFilter(aSelectedId);
+			
+			oItemsBinding.filter(oFilter);
+		},
+
+		getSuppliersFilters: function(aSelectedId) {
+			const oMultiInput = this.byId("multiInput");
+			const oModel = this.getView().getModel();
+			const aAllSuppliers = oModel.getProperty("/Suppliers");
+			const aSuppliersItems = oMultiInput.getTokens().map(el => el.getText());
+			const aFilters = [];
+			let aSuppliersId = null;
+			
+			if(aSelectedId) {
+				aSuppliersId = aSelectedId;
+			} else {
+				aSuppliersId = aAllSuppliers
+										.filter(el => aSuppliersItems.includes(el.SuppliersName))
+										.map(el => el.SupplierId);					
+			}
+			
+			aSuppliersId.forEach((el) => {
+				aFilters.push(
+					new Filter({
+						path: "Supplier",
+						operator: FilterOperator.EQ,
+						value1: el,
+						test: (supplier) => {
+							const aResult = supplier.filter((item) => {
+							
+								return	item === el
+							});
+
+							return !!aResult.length;
+						}
+					})
+				)
+			})
+
+			return aFilters.length ? aFilters : [new Filter("Supplier/0", FilterOperator.Contains, "")];
+		},
+
+		getCategoriesFilters: function() {
+			const oModel = this.getView().getModel();
+			const aCategories = oModel.getProperty("/Categories");
+			const oCategorySelect = this.byId("categorySelect");			
+			const aCategoryNames = oCategorySelect.getSelectedItems().map(el => el.getText());
+			const aCategoriesIds = aCategories
+									.filter(el => aCategoryNames.includes(el.Name))
+									.map(el => el.Id);
+			
+			const aFilters = aCategoriesIds.map((el) => {
+				return new Filter("Category", FilterOperator.Contains, el)
+			})
+			
+			return aFilters.length ? aFilters : [new Filter("Category", FilterOperator.Contains, "")];
+		},
+
+		handleValueHelpRequest: function(event) {
+			const oView = this.getView();
+			
+			if(!this.oDialog) {
+				Fragment.load({
+					id: oView.getId(),
+					name: "products.app.view.fragments.SuppliersDialog",
+					controller: this
+				}).then((oDialog) => {
+					this.oDialog = oDialog
+					oView.addDependent(this.oDialog);
+					
+					this.oDialog.open();
+				})
+			} else {
+				const oMultiInput = this.byId("multiInput");
+				const aDialogList = this.byId("valueHelpDialog");
+				const aSelectedTokens = oMultiInput.getTokens().map(el => el.getText());
+				aDialogList.getItems().map(el => el.setSelected(false));
```

Comment:
This code works but it neglects MVC principles.
You should not change the state of controls in controller. The state of control should be controlled by data (model).

This is very important because if you always change control's state in controller, your code will grow very fast, it is going to be difficult to debug, refactor, read, maintain and change the code base.

Please keep tokens data in model and use them in view.

something like that:

```xml
<StandardListItem
			...

			selected="{= ${ui/tokenData}.includes(${Id})}"

			... />
```